### ISSUE TITLE
The BaseEntity class

### ISSUE THREAD

**Issue by amalsharief** [pdevito3/craftsman#42]
Hi, I could not override the primary key Id property of the BaseEntity class. 

Is there a way to use our natural primary keys when available.  Or better yet, can you change it to BaseEntity<T>?

Best regards,

Abdoessalam

**Comment by pdevito3** [comment]
Could you describe what you're talking about here a bit more? Reading this as making the `BaseEntity` abstract with a virtual `Id` field or are you getting at something different?

**Comment by rich-mond** [comment]
It's conceivable that the Id of an entity may not be a Guid, something like an int or long.

To accommodate that, the base entity could be an abstract generic passing in the type from the property on the entity with IsPrimaryKey: true.

`public abstract class BaseEntity<T> { public T Id { get; set; } }`

**Comment by pdevito3** [comment]
Yeah this is what I went back and forth on a lot when I made this. 

On one side of things, having the convention of always having a guid PK named `Id` makes things a lot easier from a maintenance perspective and from a UX perspective for the scaffolding. I'd argue it's also a generally better practice than an int or a long but that's far from gospel. 

On the other side of things, I get how opinionated this is and it's a lot to force on devs, but I did want to bring some convention to the project and figured it was easy enough to override to an int or long and make minor updates to tests as needed if your project really called for it. 

Thoughts?

**Comment by rich-mond** [comment]
For a new project, I agree Guid would be a good choice for the primary keys. 

For the use case where there may already be an existing database where the wrapt framework sits on top providing the vertical slice, it may be necessary to accommodate different primary key types.

Opinionated by default is good, flexibility makes it even more useful 🙂

Maybe if no primary key is specified, craftsman provides a Guid Id for entities (using the generic abstract passing in a Guid) with the option of overriding the type via property in the yaml.

**Comment by pdevito3** [comment]
Yup, I get it. It does add a lot of overhead, but I'll take another look at it and see what I can figure out. 

Worst case it's easy enough to update the base entity after the fact, but it would be nice to have the added flexibility.
